packages/storage-service/src/InMemoryStorageAdapter.ts

```diff
@@ -0,0 +1,119 @@
+import type { Json } from '@metamask/utils';
+
+import type { StorageAdapter, StorageGetResult } from './types';
+import { STORAGE_KEY_PREFIX } from './types';
+
+/**
+ * In-memory storage adapter (default fallback).
+ * Implements the {@link StorageAdapter} interface using a Map.
+ *
+ * ⚠️ **Warning**: Data is NOT persisted - lost on restart.
+ *
+ * **Suitable for:**
+ * - Testing (isolated, no mocking needed)
+ * - Development (quick start, zero config)
+ * - Temporary/ephemeral data
+ *
+ * **Not suitable for:**
+ * - Production (unless data is truly ephemeral)
+ * - Data that needs to persist across restarts
+ *
+ * @example
+ * ```typescript
+ * const adapter = new InMemoryStorageAdapter();
+ * await adapter.setItem('SnapController', 'snap-id:sourceCode', 'const x = 1;');
+ * const value = await adapter.getItem('SnapController', 'snap-id:sourceCode'); // 'const x = 1;'
+ * // After restart: data is lost
+ * ```
+ */
+export class InMemoryStorageAdapter implements StorageAdapter {
+  // Explicitly implement StorageAdapter interface
+  /**
+   * Internal storage map.
+   */
+  readonly #storage: Map<string, string>;
+
+  /**
+   * Constructs a new InMemoryStorageAdapter.
+   */
+  constructor() {
+    this.#storage = new Map();
+  }
+
+  /**
+   * Retrieve an item from in-memory storage.
+   * Deserializes JSON data from storage.
+   *
+   * @param namespace - The controller namespace.
+   * @param key - The data key.
+   * @returns StorageGetResult: { result } if found, {} if not found, { error } on failure.
+   */
+  async getItem(namespace: string, key: string): Promise<StorageGetResult> {
+    const fullKey = `${STORAGE_KEY_PREFIX}${namespace}:${key}`;
+    const serialized = this.#storage.get(fullKey);
+
+    // Key not found - return empty object
+    if (serialized === undefined) {
+      return {};
+    }
+
+    try {
+      const result = JSON.parse(serialized);
+      return { result };
+    } catch (error) {
+      console.error(`Failed to parse stored data for ${fullKey}:`, error);
+      return { error: error as Error };
```

Comment:
Normally I don't like to cast if we can avoid it, and we _can_ avoid it here by validating the error type.

But... it _is_ pretty easy to tell in this case that the only possible error here is `SyntaxError`. So.... this seems pretty unobjectionable.

I might propose an improvement here later, but I can do it as a separate PR